src/QsCompiler/SyntaxProcessor/CapabilityInference/ResultAnalyzer.fs

```diff
@@ -0,0 +1,186 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+module Microsoft.Quantum.QsCompiler.SyntaxProcessing.CapabilityInference.ResultAnalyzer
+
+open Microsoft.Quantum.QsCompiler
+open Microsoft.Quantum.QsCompiler.DataTypes
+open Microsoft.Quantum.QsCompiler.Diagnostics
+open Microsoft.Quantum.QsCompiler.SyntaxProcessing.CapabilityInference.ContextRef
+open Microsoft.Quantum.QsCompiler.SyntaxTokens
+open Microsoft.Quantum.QsCompiler.SyntaxTree
+open Microsoft.Quantum.QsCompiler.Transformations.Core
+
+type ResultDependency =
+    /// An equality expression inside the condition of an if statement, where the operands are results
+    | ConditionalEqualityInOperation
+
+    /// An equality expression outside the condition of an if statement, where the operands are results.
+    | UnrestrictedEquality
+
+    /// A return statement in the branch of an if statement whose condition depends on a result.
+    | ReturnInDependentBranch
+
+    /// A set statement in the branch of an if statement whose condition depends on a result, but which is reassigning a
+    /// variable declared outside the block.
+    | SetInDependentBranch of name: string
+
+type Context =
+    {
+        CallableKind: QsCallableKind
+        FrozenVars: string Set
+        InCondition: bool
+    }
+
+let createPattern kind range =
+    let opacity =
+        match kind with
+        | ConditionalEqualityInOperation -> ResultOpacity.controlled
+        | UnrestrictedEquality
+        | ReturnInDependentBranch
+        | SetInDependentBranch _ -> ResultOpacity.transparent
+
+    let capability = TargetCapability.withResultOpacity opacity TargetCapability.bottom
+
+    let diagnose (target: Target) =
+        let range = QsNullable.defaultValue Range.Zero range
+
+        match kind with
+        | _ when TargetCapability.subsumes target.Capability capability -> None
+        | ConditionalEqualityInOperation
+        | UnrestrictedEquality ->
+            let code =
+                if target.Capability.ResultOpacity = ResultOpacity.opaque then
+                    ErrorCode.UnsupportedResultComparison
+                else
+                    ErrorCode.ResultComparisonNotInOperationIf
+
+            Some(code, [ target.Name ])
+        | ReturnInDependentBranch ->
+            if target.Capability.ResultOpacity = ResultOpacity.opaque then
+                None
```

Comment:
I am reading this as:
If we have a return in a branch whose execution depends on a measurement result, and the target doesn't support result comparisons anywhere, then give no error. Is that because in this case we already gave an error for the branching in the first place?